pkg/discovery/worker/controller_metrics_collector.go

```diff
@@ -0,0 +1,171 @@
+func (collector *ControllerMetricsCollector) CollectControllerMetrics() ([]*repository.KubeController, error) {
+	if collector.cluster == nil {
+		return nil, fmt.Errorf("error collecting K8s controller metrics, because cluster summary object is null for discovery worker %s",
+			collector.workerId)
+	}
+	kubeNamespaceMap := collector.cluster.NamespaceMap
+	// Map from controller UID to the corresponding kubeController
+	kubeControllersMap := make(map[string]*repository.KubeController)
+	var kubeControllerList []*repository.KubeController
+	for _, pod := range collector.podList {
+		if pod.OwnerReferences == nil {
+			// If pod has no OwnerReferences, it is a bare pod without controller. Skip this.
+			glog.V(4).Infof("Skip creating KubeController for bare Pod %s", util.PodKeyFunc(pod))
+			continue
+		}
+		controllerType, controllerName, controllerUID, err := collector.getKubeControllerInfo(metrics.PodType, util.PodKeyFunc(pod))
+		if err != nil {
+			glog.Errorf("Error getting controller info from Pod %s: %v", util.PodKeyFunc(pod), err)
+			continue
+		}
+		kubeController, exists := kubeControllersMap[controllerUID]
+		if !exists {
+			namespace := pod.Namespace
+			// Create default KubeController entity
+			kubeController = repository.NewKubeController(collector.cluster.Name, namespace, controllerName, controllerType,
+				controllerUID)
+			kubeNamespace, namespaceExists := kubeNamespaceMap[namespace]
+			if !namespaceExists {
+				glog.Errorf("Namespace %s does not exist in cluster %s", namespace, collector.cluster.Name)
+				continue
+			}
+			// Add quota resources to KubeController entity with capacity as namespace quota capacity
+			for _, resourceType := range metrics.QuotaResources {
+				namespaceQuotaResource, err := kubeNamespace.GetAllocationResource(resourceType)
+				resourceCapacity := repository.DEFAULT_METRIC_CAPACITY_VALUE
+				if err != nil {
+					glog.Errorf("KubeNamespace %s has invalid resource %s. Set %s capacity of k8s controller %s to %v",
+						kubeNamespace.Name, resourceType, resourceType, kubeController.GetFullName(), repository.DEFAULT_METRIC_CAPACITY_VALUE)
+				} else {
+					// For CPU resources, set capacity values in number of cores here. Will update CPU resources capacity
+					// to MHz based on average node CPU frequency when building workload controller entity DTO in
+					// controller_discovery_worker. Each k8s controller metrics collector collects controller metrics
+					// from certain amount of nodes run by a k8s discovery worker. So at this point, there's no way to
+					// calculate average CPU frequency of all nodes in the cluster.
+					resourceCapacity = namespaceQuotaResource.Capacity
+				}
+				kubeController.AddAllocationResource(resourceType, resourceCapacity, repository.DEFAULT_METRIC_VALUE)
+			}
+			kubeControllersMap[controllerUID] = kubeController
+			kubeControllerList = append(kubeControllerList, kubeController)
+		}
+		kubeController.Pods = append(kubeController.Pods, pod)
```

Comment:
Whats the reason of collecting these pods, I haven't seen these used so far.

Reply:
I should have added a TODO here. I plan to use the pods of each controller to create ContainerSpec entity to simplify the logic in the committed changes in `container_spec_entity_dto_builder.go` (https://github.com/marvelyue/kubeturbo/blob/14ed3c34c23a28f257059d52a093e373d418c46e/pkg/discovery/dtofactory/container_spec_entity_dto_builder.go)

ContainerSpec entity models a shared portion of container replicas defined by the K8s controller. Currently the logic in `container_spec_entity_dto_builder.go` is to get controller UID from all pods and create a map from controllerUID to container names. After creating KubeController entities with corresponding pods, we'll be able to simply the logic.

Reply:
Ack!